fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,508 @@
+                .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS)
+                .addAllTargetIndex(targetIndexes)
+                .build();
+    }
+
+    private static boolean areTheyAllIdempotent(@Nonnull FDBRecordStore store, List<Index> targetIndexes) {
+        return targetIndexes.stream()
+                .allMatch(targetIndex -> store.getIndexMaintainer(targetIndex).isIdempotent());
+    }
+
+    @Override
+    List<Object> indexingLogMessageKeyValues() {
+        return Arrays.asList(
+                LogMessageKeys.INDEXING_METHOD, "mutual multi target by records",
+                LogMessageKeys.TARGET_INDEX_NAME, common.getTargetIndexesNames()
+        );
+    }
+
+    private List<Tuple> getPrimaryKeyBoundaries(@Nonnull FDBRecordStore store) {
+        TupleRange tupleRange = common.computeRecordsRange();
+        store.getContext().getReadVersion(); // for instrumentation reasons
+        List<Tuple> boundaries;
+        try (RecordCursor<Tuple> cursor = store.getPrimaryKeyBoundaries(tupleRange)) {
+            boundaries = cursor.asList().join();
+        }
+
+        if (boundaries == null) {
+            boundaries = new ArrayList<>();
+        }
+
+        // Add the two endpoints to the range
+        if (tupleRange == null) {
+            // Here: make sure that the boundaries cover everything
+            boundaries.add(0, null);
+            boundaries.add(null);
+        } else {
+            // Here: add the endpoints, unless they are already included
+            if (boundaries.isEmpty() || tupleRange.getLow() == null || tupleRange.getLow().compareTo(boundaries.get(0)) < 0) {
+                boundaries.add(0, tupleRange.getLow());
+            }
+            if (tupleRange.getHigh() == null || tupleRange.getHigh().compareTo(boundaries.get(boundaries.size() - 1)) > 0) {
+                boundaries.add(tupleRange.getHigh());
+            }
+        }
+
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info(KeyValueLogMessage.of("got boundaries",
+                    LogMessageKeys.INDEX_NAME, common.getTargetIndexesNames(),
+                    LogMessageKeys.RANGE, tupleRange,
+                    LogMessageKeys.KEY_COUNT, boundaries.size()));
+        }
+
+        return boundaries;
+    }
+
+    private int getPrimeStep(int size, Random rn) {
+        // Find a prime step that isn't size's factor
+        validateOrThrowEx(size > 0, "No ranges to build");
+        if (size < 3) {
+            return 1;
+        }
+        for (int sanity = 0; sanity < 800; sanity ++) {
+            int step = rn.nextInt(size);
+            if ((step & 1) == 0) {
+                step --;
+            }
+            if (step < 2) {
+                return 1;
+            }
+            // Note: isPrime is @Beta, but this code should work even if it's wrong. The real protection is in "size % step".
+            if ((size % step != 0) && IntMath.isPrime(step)) {
```

Comment:
You are correct. But I rather keep it a prime to reduce the change of two indexing process using steps with common factors - which may lead to some inefficient correlation. (It could have been avoided with `IntMath.gcd(stepOfProcessA, stepOfProcessB)`, but that to hard to implement ..)

Reply:
I will update the comment, though.